capa/features/extractors/vmray/models.py

```diff
@@ -0,0 +1,334 @@
+    "signed_16bit",
+    "unsigned_16bit",
+    "signed_32bit",
+    "unsigned_32bit",
+    "signed_64bit",
+    "unsigned_64bit",
+    "double",
+    "void_ptr",
+    "bool",
+    "unknown",
+    "ptr",
+    "void",
+    "str",
+    "array",
+    "container",
+    "bindata",
+    "undefined_type",
+)
+"""
+
+PARAM_TYPE_PTR = ("void_ptr", "ptr")
+PARAM_TYPE_STR = ("str",)
+PARAM_TYPE_INT = (
+    "signed_8bit",
+    "unsigned_8bit",
+    "signed_16bit",
+    "unsigned_16bit",
+    "signed_32bit",
+    "unsigned_32bit",
+    "signed_64bit",
+    "unsigned_64bit",
+    "double",
+    "bool",
+    "unknown",
+)
+
+
+def xml_to_dict(xml):
+    return xmltodict.parse(xml, attr_prefix="")
+
+
+def hexint(value: Union[str, int]) -> int:
+    if isinstance(value, str):
+        return int(value, 16) if value.startswith("0x") else int(value, 10)
+    else:
+        return value
+
+
+def validate_hex_int(value: Union[str, int]) -> int:
+    return hexint(value)
+
+
+# convert the input value to a Python int type before inner validation (int) is called
+HexInt = Annotated[int, BeforeValidator(validate_hex_int)]
+
+
+# models flog.xml file
+class ParamDeref(BaseModel):
+    type_: str = Field(alias="type")
+    value: Optional[str] = None
+
+
+class Param(BaseModel):
+    name: str
+    type_: str = Field(alias="type")
+    value: Optional[str] = None
+    deref: Optional[ParamDeref] = None
+
+
+def validate_param_list(value: Union[List[Param], Param]) -> List[Param]:
+    if isinstance(value, list):
+        return value
+    else:
+        return [value]
+
+
+# params may be stored as a list of Param or a single Param so we convert
+# the input value to Python list type before the inner validation (List[Param])
+# is called
+ParamList = Annotated[List[Param], BeforeValidator(validate_param_list)]
+
+
+class Params(BaseModel):
+    params: ParamList = Field(alias="param")
+
+
+def validate_call_name(value: str) -> str:
+    if value.startswith("sys_"):
+        # VMRay appears to log kernel function calls ("sys_*") for Linux so we remove that
+        # here to enable capa matching
+        return value[4:]
+    else:
+        return value
+
+
+# function call names may need to be reformatted to remove data, etc. so we reformat
+# before calling the inner validation (str)
+CallName = Annotated[str, BeforeValidator(validate_call_name)]
+
+
+class FunctionCall(BaseModel):
+    # ts: HexInt
+    fncall_id: HexInt
+    process_id: HexInt
+    thread_id: HexInt
+    name: CallName
+    # addr: HexInt
+    # from_addr: HexInt = Field(alias="from")
+    params_in: Optional[Params] = Field(alias="in", default=None)
+    params_out: Optional[Params] = Field(alias="out", default=None)
+
+
+class FunctionReturn(BaseModel):
+    ts: HexInt
+    fncall_id: HexInt
+    addr: HexInt
+    from_addr: HexInt = Field(alias="from")
+
+
+class Analysis(BaseModel):
+    log_version: str  # tested 2
+    analyzer_version: str  # tested 2024.2.1
+    # analysis_date: str
```

Comment:
ATM I'm not sure where this information would be used so I think it should remain a comment for now. Happy to discuss more if you have a specific place you think it could be used, e.g. as part of VMRay-related error messages?

Reply:
You're right. I think it would be useful but we're not tracking it right now for dynamic analysis. We can come back to it when needed.